drivers: nrf: Rework UARTE shim for uart ASYNC API and add power management

Mierunski opened the pull request:
Rework uart_nrfx_uarte shim to work with asynchronous API.

Fix power management issues

Fixes #12501

Mierunski committed:
drivers: nrf: Rework UARTE shim for uart ASYNC API.

Rework uart_nrfx_uarte shim to work with asynchronous API.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
dts: nrf: Add DPPIC to device tree.

Add DPPIC to dts. Add HAS_HW_NRF_DPPIC to nrf91 soc.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski commented:
@jarz-nordic @nordic-krch @anangl ping

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -12,10 +12,19 @@
+#if CONFIG_HAS_HW_NRF_PPI
+#include "nrfx_ppi.h"
+#else
+#include "nrfx_dppi.h"
```
I would suggest `elif` and `else error`

jarz-nordic commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -69,6 +76,17 @@ config UART_0_NRF_TX_BUFFER_SIZE
+	depends on UART_0_NRF_UARTE
+	depends on UART_ASYNC_API
+	help
+	  Hardware RX byte counting requires timer instance and one PPI channel
```
Please elaborate a bit more why it is recommended.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	}
+}
+
+static void rx_timeout(struct device *dev)
```
it would be good to explain that mechanism in the comment. now it is quite unclear.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	u32_t ret;
+
+	if (IS_ENABLED(CONFIG_UARTE_NRF_HW_ASYNC) && data->async->hw_async) {
+		ret = nrfx_ppi_channel_alloc(&data->async->rx_cnt.ppi);
```
what about dppi? maybe it would be worth extracting dppi/ppi stuff to another file with simple call like `ppi_configure` or `gppi_configure` where g stands for generic :)

Mierunski replied:
There are some differences in how PPI and DPPI handle signals, I believe it can be done but it won't be a part of this PR

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+			LOG_ERR("Failed to allocate PPI Channel, "
+					"switching to software byte counting.");
+			data->async->hw_async = false;
+			data->async->rx_cnt.cnt = 0;
```
return error

Mierunski replied:
If we return error here, device won't be initialized and if this UARTE instance is set as shell output, user won't see any information what has gone wrong.

nordic-krch reviewed (changes requested):
im leaving some initial comments but i'm not finished with this! will come back soon.

anangl reviewed (commented):
A couple of minor comments for the beginning. I just started digging in the code (well, some comments there could help a lot...).

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+			data->async->rx_cnt.cnt = 0;
+		}
+		nrfx_timer_enable(&data->async->timer);
+		nrfx_timer_clear(&data->async->timer);
```
These two calls should be in `else {}` block.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	return 0;
+}
+
+#define INTERRUPTS_MASK							       \
```
What is the gain of having it defined here and not using the masks directly in the only place in the code where they are needed?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	nrf_uarte_tx_buffer_set(cfg->uarte_regs, buf, len);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTARTED);
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTTX);
+	if (cfg->uarte_regs->CONFIG & UARTE_CONFIG_HWFC_Msk) {
```
Why not `if (cfg->flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)`?
We avoid operating directly on the hardware registers in any layers upper than HAL. If the HAL is missing some required functionality, it needs to be extended.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+			return -EIO;
+		}
+#else
+		cfg->uarte_regs->PUBLISH_RXDRDY =
```
Use HAL for this (`nrf_uarte_publish_set`).

anangl reviewed (commented):
Some more comments. Not finished reviewing.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	user_callback(dev, &evt);
+	data->async->tx = NULL;
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTOPPED);
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STOPTX);
```
Won't this result in `UART_TX_ABORTED` event notification? I guess `data->async->tx` should be checked in `txstopped_isr` to avoid this.

But wouldn't it be better to trigger this task before calling `user_callback` (and to change the order of event handling in `uarte_nrfx_isr` so that the `TXSTOPPED` event could be handled right after the `ENDTX` one, in the same ISR run)?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	};
+	user_callback(dev, &evt);
+	data->async->tx = NULL;
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTOPPED);
```
This event is cleared in `uarte_nrfx_isr` so this is redundant.

Mierunski commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -69,6 +76,17 @@ config UART_0_NRF_TX_BUFFER_SIZE
+	  Hardware RX byte counting requires timer instance and one PPI channel
+
+config UART_0_NRF_HW_ASYNC_TIMER
+	int "Timer instance"
```
There is no default value, user has to set it explicitly

jarz-nordic replied:
So as for now default value is `n`

ioannisg replied:
If I understand kconfig properly (@ulfalizer can help) this will be a bit undefined, i.e. it's not really a "not-set", cause the value is boolean. So, it appears to me that the macro is still generated (Without a value though)

Mierunski replied:
If value is not set then cmake will report an error and won't generate the project

ioannisg replied:
We should give the developer the option to build this without menuconfig, and as far as I can understand Kconfig, this requires a default value for int- Kconfig symbols.

Mierunski replied:
This value is taken into account only when developer selects hardware byte counting. When build with default configuration there are no issues with it. I don't want to select a default value here, as it should be explicitly set by developer if he decides to use hardware byte counting.

ioannisg replied:
Understood, now, thanks. So, I suggest you do as @ulfalizer has done in several similar cases, i.e. remove the "depends on" and add the whole Kconfig definition in i`f UART_0_NRF_HW_ASYNC ... endif #UART_0_NRF_HW_ASYNC`

ulfalizer replied:
There's some notes on `int` defaults on the [best practices page](https://docs.zephyrproject.org/latest/guides/kconfig/index.html#redundant-defaults).

Usually, it's a good idea to give `int` symbols defaults, because they don't default to zero (for semi-messy reasons). If no default is given, then the `prj.conf` (or board `defconfig` files, etc.) must set the symbol. Otherwise, you'll get a `.config` that generates warnings when read back in.

(Super nit: Trying to standardize making it `# UART_0_NRF_HW_ASYNC` instead of `#UART_0_NRF_HW_ASYNC` too btw. It's more common.)

ulfalizer replied:
Re. `if` vs `depends on`, `depends on` is cleaner if there's just a single symbol with the dependency.

If there's many symbols with the same dependency in a row, `if` is cleaner.

It's a bit subjective when you have just two symbols or the like. It's good to be aware that `if FOO` is just a shorthand for adding `depends on FOO` to all the stuff within at least.

People tend to read a bunch of things into Kconfig that aren't there (probably because the official docs aren't great). :)

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	u32_t read;
+
+	if (IS_ENABLED(CONFIG_UARTE_NRF_HW_ASYNC) && data->async->hw_async) {
```
This condition should be wrapped in an `inline` function (e.g. `hw_rx_counting_enabled`) and called wherever there is a need to check if the hardware counting is active. This way all the parts of code that appear to be dead have chance to be optimized out. Right now this won't happen, for example, to the `if` block related to the `RXDRDY` event in `uarte_nrfx_isr`.

BTW `hw_async` is very cryptic.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	data->async->tx_len = len;
+	data->async->tx_cur_idx = 0;
+	nrf_uarte_tx_buffer_set(cfg->uarte_regs, buf, len);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTARTED);
```
This event does not seem to be used anywhere. Why clear it?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -623,189 +1132,186 @@ static int uarte_instance_init(struct device *dev,
 
-#define UARTE_CONFIG_INT(idx)				\
-	.tx_buff_size = UARTE_TX_BUFFER_SIZE(idx)
+#define UARTE_INT_DRIVEN(idx)						       \
```
What about using instead something like this:
```
#define UARTE_INT_DRIVEN(idx)						       \
	COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN),	       \
		(static u8_t uarte##idx##_tx_buffer[			       \
			max(CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE,	       \
			    BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE))];      \
		 static struct uarte_nrfx_int_driven			       \
		 	uarte##idx##_int_driven = {			       \
				.tx_buffer = uarte##idx##_tx_buffer,	       \
				.tx_buff_size = sizeof(uarte##idx##_tx_buffer),\
			};),						       \
		())
```
Then you could add just: `UARTE_INT_DRIVEN(idx)` to `UART_NRF_UARTE_DEVICE(idx)` and initialize `uarte_##idx##_data` this way:
```
	static struct uarte_nrfx_data uarte_##idx##_data = {		       \
		UARTE_CONFIG(idx),					       \
		COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN),  \
			(.int_driven = &uarte##idx##_int_driven),	       \
			()),						       \
	}
```
Same could be done for `ASYNC` stuff and you'll no longer need to use `#ifdef CONFIG_UART_x_ASYNC` and `#ifdef CONFIG_UART_x_INTERRUPT_DRIVEN` for every instance.
Additionally, the fields `int_driven` and `async` won't need to be always present in `uarte_nrfx_data`.

Mierunski committed:
tests: uart: Add chained write test case for UART async API

Add new test case

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
tests: uart: Add nrf9160_pca10090 to UART async test

Add nrf9160_pca10090 to UART async test.
Add myself as codeowner to uart_async_api tests.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -20,19 +42,65 @@
+/* Device data structure */
+struct uarte_nrfx_data {
+	struct uart_config uart_config;
+	struct uarte_nrfx_int_driven *int_driven;
```
Wrap in `#ifdef UARTE_INTERRUPT_DRIVEN`.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -20,19 +42,65 @@
+struct uarte_nrfx_data {
+	struct uart_config uart_config;
+	struct uarte_nrfx_int_driven *int_driven;
+	struct uarte_async_cb *async;
```
Wrap in `#ifdef CONFIG_UART_ASYNC_API`.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -550,69 +1094,71 @@ static int uarte_instance_init(struct device *dev,
 	}
-#endif /* UARTE_INTERRUPT_DRIVEN */
+
+#if IS_ENABLED(CONFIG_UART_ASYNC_API)
```
`#ifdef CONFIG_UART_ASYNC_API`

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -550,69 +1094,71 @@ static int uarte_instance_init(struct device *dev,
-		return uarte_instance_init(dev,				       \
-				&init_config,				       \
-				UARTE_##idx##_INTERRUPT_DRIVEN);	       \
+		COND_CODE_1(UART_ISR_ENABLED,				       \
```
This will cause that interrupts will be connected and enabled for all UARTE instances if at least one of them will be using interrupt driven or asynchronous API. I guess it was not the intention.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -286,8 +820,24 @@ static int uarte_nrfx_poll_in(struct device *dev, unsigned char *c)
+		uarte_nrfx_tx(dev, &c, 1, 100);
+		while (!finished) {
+		}
+		uarte_nrfx_callback_set(dev, prev_cb, usr_context);
```
This hack is really ugly. Especially with this hard-coded timeout. Please, use something nicer, for example, disable the `ENDTX` interrupt here or use a different handling of the `ENDTX` event in `uarte_nrfx_isr` if `data->async->tx` is `NULL`.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+				      NRF_UARTE_INT_RXDRDY_MASK);
+	}
+
+
```
Extra empty line.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	}
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STOPTX);
+
+	if (cfg->uarte_regs->CONFIG & UARTE_CONFIG_HWFC_Msk) {
```
`if (data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS) {`

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -11,6 +11,28 @@
+#include <logging/log.h>
+
+/* Generalize PPI or DPPI channel management */
+#if CONFIG_HAS_HW_NRF_PPI
```
`#if defined(CONFIG_HAS_HW_NRF_PPI)` here
and
`#elif defined(CONFIG_HAS_HW_NRF_DPPI)` below.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	 : 0)
+
+static void error_isr(struct device *dev)
+{
```
The reception should be stopped in this handler. It is not done automatically by hardware when the `ERROR` event occurs.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+}
+
+static void rxto_isr(struct device *dev)
+{
```
It would be good to add a comment that this handler is called when the reception is interrupted, in contrary to finishing the reception after filling all provided buffers, in which case the events `UART_RX_BUF_RELEASED` and `UART_RX_DISABLED` are reported from `endrx_isr`. It's not that obvious.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	data->async->rx_total_user_byte_cnt = data->async->rx_total_byte_cnt;
+	data->async->rx_next_size = 0;
+	data->async->rx_next_buf = NULL;
+	data->async->rx_offset = 0;
```
This should be done in `uarte_nrfx_rx_enable` instead. Currently the offset will not be cleared when the reception is disabled, for instance.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXSTARTED);
+	if (!hw_rx_counting_enabled(data)) {
+		nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXDRDY);
```
No need to do it here. `uarte_nrfx_isr` takes care of this.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	data->async->rx_total_byte_cnt += evt.data.rx.len;
+	evt.type = UART_RX_BUF_RELEASED;
+	evt.data.rx_buf.buf = data->async->rx_buf;
+	user_callback(dev, &evt);
```
What if the user decides to reuse the released buffer and provides it in a call to `uart_rx_buf_rsp` from the event handler? API does not seem to forbid such scenario.

Mierunski replied:
`uart_rx_buf_rsp` can be called only in response to `UART_RX_BUF_REQUEST`

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+
+static void tx_timeout(struct device *dev)
+{
+	uarte_nrfx_tx_abort(dev);
```
The return value is ignored here. Some tools may complain about that.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -286,8 +820,24 @@ static int uarte_nrfx_poll_in(struct device *dev, unsigned char *c)
 static void uarte_nrfx_poll_out(struct device *dev,
```
What about `uarte_nrfx_poll_in`? Currently it will always return `-1` when `CONFIG_UART_ASYNC_API` is enabled.

Mierunski replied:
There is no possibility to handle poll_in while using asynchronous API, I will change it to return -EBUSY

Mierunski committed:
boards: nrf: Enable TIMER1 by default in nrf9160_pca10090

Enable TIMER1 by default in nrf9160_pca10090 so it can be selected
by user for hardware byte counting in UARTE or other purpose.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
drivers: nrf: Enable usage of Async UART API for UARTE on nrf91

Modify driver to use PPI or DPPI depending on soc.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	user_callback(dev, &evt);
+	data->async->rx_buf = NULL;
+	if (data->async->rx_next_buf) {
+		evt.data.rx_buf.buf = data->async->rx_next_buf;
```
Good point, I will patch it later

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +321,468 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	}
+}
+
+static void timer_handler(nrf_timer_event_t event_type, void *p_context) { }
```
This could be updated in `nrfx` ;)

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +321,468 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	data->async->tx = buf;
+	data->async->tx_len = len;
+	data->async->tx_cur_idx = 0;
+	nrf_uarte_tx_buffer_set(cfg->uarte_regs, buf, len);
```
I recommend to check if data is in RAM

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +321,468 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	}
+
+	/* Check if data was received since last function call */
+	if (read > data->async->rx_total_byte_cnt) {
```
Wouldn't it be safer to use `!=`? In theory, when a continuous receiving is enabled with next buffers provided consecutively, the total counter could wrap around.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -316,25 +866,29 @@ static void uarte_nrfx_poll_out(struct device *dev,
 	}
 
+	/* reset transmitter ready state */
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDTX);
```
This will break the `INTERRUPT_DRIVEN` mode.

Mierunski commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -283,11 +822,25 @@ static int uarte_nrfx_poll_in(struct device *dev, unsigned char *c)
+
+	if (data->async) {
+		while (data->async->tx) {
+		}
```
Added volatile

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -283,11 +820,25 @@ static int uarte_nrfx_poll_in(struct device *dev, unsigned char *c)
+		/* Simulate tx buffer present,
+		 * this way uarte_nrfx_tx will return -EBUSY.
+		 */
+		data->async->tx_buf = &c;
```
`uarte_nrfx_tx` will return `-EBUSY` but another `uarte_nrfx_poll_out` will just hang in the above loop. See the long comment right below for more context.

Mierunski commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -623,189 +1197,102 @@ static int uarte_instance_init(struct device *dev,
+
+#define UARTE_INT_DRIVEN(idx)						       \
+	COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN),	       \
+		(static u8_t uarte##idx##_tx_buffer[\
```
I can't, checkpatch is complaining about it

anangl replied:
Complaining? Why would it? Maybe you used 8 spaces instead of a tab somewhere?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +323,476 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+			data->async->rx_total_user_byte_cnt =
+				data->async->rx_total_byte_cnt;
+			user_callback(dev, &evt);
+			data->async->rx_timeout_left = data->async->rx_timeout;
```
This seems to be redundant. You can't enter this block again without first getting inside `if (read != data->async->rx_total_byte_cnt) {` above.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -20,19 +41,73 @@
+	u8_t *rx_next_buf;
+	u32_t rx_total_byte_cnt; /* Total number of bytes received */
+	u32_t rx_total_user_byte_cnt; /* Total number of bytes passed to user */
+	nrfx_timer_t timer;
```
This should be moved to `struct uarte_nrfx_config`.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +323,476 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_RXSTARTED);
+
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+	data->async->rx_enabled = true;
```
Move this before triggering the `STARTRX` task. Otherwise, you may end up servicing `endrx_isr` without this field being set.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -20,19 +41,73 @@
+	struct k_timer tx_timeout_timer;
+
+	u8_t *rx_buf;
+	size_t rx_size;
```
This field is not used, it is only assigned a value in a few places, and one of them is `rx_next_size` which in turn is never assigned any value (it stays initialized by default to `0`). Both these fields should be removed.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +323,472 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+
+	if (nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_RXSTARTED)) {
+		nrf_uarte_event_clear(uarte,
+				      NRF_UARTE_EVENT_RXSTARTED);
```
Nit: this line doesn't seem to need this break.
Same story below, with `NRF_UARTE_EVENT_TXSTOPPED`.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -20,19 +41,70 @@
 #endif
 
+#define MS_PER_TICK (CONFIG_SYS_CLOCK_TICKS_PER_SEC <= 1000 ?	\
+		     1000 / CONFIG_SYS_CLOCK_TICKS_PER_SEC : 1)
```
This integer division should be rather rounded up. You can use e.g. `NRFX_CEIL_DIV` for this.
But actually, do we still need this value of `MS_PER_TICK`? It is used only in calculation of `rx_timeout_slab`. Won't the effect be the same if you use just `1` there?

anangl reviewed (approved):
LGTM, finally. :wink:
Left only some minor comments.

nordic-krch reviewed (commented):
I have some minor comments.

Mierunski commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +321,471 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+		}
+	}
+
+	if (hw_rx_counting_enabled(data)) {
```
No, in current way if hardware counting initialization fails at any stage, driver is still initialized.

Mierunski commented:
@carlescufi Could you merge this?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +321,487 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	u32_t ret;
```
`int ret`

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -251,6 +321,487 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	u32_t ret = uarte_nrfx_rx_counting_init(dev);
```
`int ret`

carlescufi commented:
@Mierunski we are currently in 1.14 release mode, so unless it's a bugfix it will remain there until we release. Is this considered a bugfix? Is it fixing a particular bug?

Mierunski committed:
drivers: nrf: Add power management to uarte shim

Adds power management to uarte shim.

Fixes 12501

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski commented:
Fixes #12501

ioannisg commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_common.dts:
```diff
@@ -121,6 +121,10 @@
 	status = "ok";
 };
 
+&timer1 {
```
Why does it need to be enabled? Can't see from commit message :)

Mierunski replied:
So it can be selected by user for hardware byte counting. Other timers are enabled by default. Also in other socs all timers are enabled. Just following the pattern here

ioannisg commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,6 +79,20 @@ config UART_0_NRF_TX_BUFFER_SIZE
+	depends on UART_0_NRF_UARTE
+	depends on UART_ASYNC_API
+	help
+	  In default driver uses interrupts to count incoming bytes, it is possible
```
I suppose "In" --> "If")?

ioannisg replied:
Same for all similar Kconfig options

Mierunski replied:
fixed

ioannisg commented on dts/arm/nordic/nrf9160_common.dtsi:
```diff
@@ -29,6 +29,13 @@ adc: adc@e000 {
 	label = "ADC_0";
 };
 
+dppi: dppi@17000 {
```
I wonder; why didn't you name the node dppic, as its proper name (also in MDK headers)?

Mierunski replied:
changed to dppic

ioannisg reviewed (commented):
Some comments/suggestions from my side

anangl commented on ext/hal/nordic/Kconfig:
```diff
@@ -19,6 +19,9 @@ config NRFX_ADC
 config NRFX_CLOCK
 	bool
 
+config NRFX_DPPIC
```
While renaming the DT node to `dppic` is ok, because it refers to the DPPIC peripheral, this option should stay with the name `NRFX_DPPI`, since it enables nrfx driver `nrfx_dppi` (without the `c`).

ulfalizer commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -198,8 +267,70 @@ config UART_3_NRF_TX_BUFFER_SIZE
+if UART_0_NRF_HW_ASYNC_TIMER = 0 || UART_1_NRF_HW_ASYNC_TIMER = 0 || \
+	UART_3_NRF_HW_ASYNC_TIMER = 0 || UART_2_NRF_HW_ASYNC_TIMER = 0
+config NRFX_TIMER0
+	default y
```
Could use a `depends on` on the symbol instead of an `if` here.

`default y if UART_0_HW_ASYNC_TIMER = 0 || ...` will work too.

ulfalizer commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -198,8 +267,70 @@ config UART_3_NRF_TX_BUFFER_SIZE
+if UART_0_NRF_HW_ASYNC_TIMER = 1 || UART_1_NRF_HW_ASYNC_TIMER = 1 || \
+	UART_3_NRF_HW_ASYNC_TIMER = 1 || UART_2_NRF_HW_ASYNC_TIMER = 1
+config NRFX_TIMER1
+	default y
```
Could use `depends on`.

ulfalizer commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -198,8 +267,70 @@ config UART_3_NRF_TX_BUFFER_SIZE
+if UART_0_NRF_HW_ASYNC_TIMER = 2 || UART_1_NRF_HW_ASYNC_TIMER = 2 || \
+	UART_3_NRF_HW_ASYNC_TIMER = 2 || UART_2_NRF_HW_ASYNC_TIMER = 2
+config NRFX_TIMER2
+	default y
```
Ditto here and below.

ulfalizer reviewed (approved):
Kconfig nits fixed
